src/pubsub/ua_pubsub_reader.c

```diff
@@ -788,15 +806,27 @@ UA_DataSetReader *UA_ReaderGroup_findDSRbyId(UA_Server *server, UA_NodeId identi
 void UA_ReaderGroup_subscribeCallback(UA_Server *server, UA_ReaderGroup *readerGroup) {
     UA_PubSubConnection *connection =
         UA_PubSubConnection_findConnectionbyId(server, readerGroup->linkedConnection);
+    if (!connection) {
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER, "SubscribeCallback(): "
+            "Find linked connection failed");
+        UA_ReaderGroup_setPubSubState(server, UA_PUBSUBSTATE_ERROR, readerGroup);
+        return;
+    }
     UA_ByteString buffer;
-
     if(UA_ByteString_allocBuffer(&buffer, 4096) != UA_STATUSCODE_GOOD) {
-        UA_LOG_INFO(&server->config.logger, UA_LOGCATEGORY_SERVER, "Message buffer alloc failed!");
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER, "SubscribeCallback(): Message buffer alloc failed!");
+        UA_ReaderGroup_setPubSubState(server, UA_PUBSUBSTATE_ERROR, readerGroup);
         return;
     }
 
+    UA_StatusCode res = connection->channel->receive(connection->channel, &buffer, NULL, 1000);
+    if ((res & 0x80000000) != 0) { // status is bad
```

Comment:
It's not good to read. Why not just !UA_STATUSCODE_GOOD?

Reply:
I think the receive function of the UDP plugin for instance returns UA_STATUSCODE_GOODNONCRITICALTIMEOUT and sets the message buffer length to 0 if there are no new packages. https://github.com/open62541/open62541/blob/9fab2726e510f9cab2123047a24b6c401a12a2a6/plugins/ua_pubsub_udp.c#L485

So the returned status code is not UA_STATUSCODE_GOOD, but UA_STATUSCODE_GOODNONCRITICALTIMEOUT. In this case I do not want to print a error message or set the state to error. 
Only if a bad statuscode gets returned.
That's why I have masked the status code, because I think that all Bad codes set their highest bit to '1'

Reply:
I would avoid the bitfield magic.

Can you do an explicit comparison against UA_STATUSCODE_GOODNONCRITICALTIMEOUT?

Or we can add `UA_StatusCode_isBad` to make it explicit.

Reply:
Sure, I think I will implement the UA_StatusCode_isBad function, if that's ok.
Because the ua_pubsub_ethernet_etf.c receive implementation can return UA_STATUSCODE_GOODNODATA or UA_STATUSCODE_GOOD or UA_STATUSCODE_GOODNONCRITICALTIMEOUT and then the if() would look more complex

Reply:
Where may I implement it?
ua_util.c and .h?

Reply:
I have added it to ua_util

Reply:
You can add it to types.h / ua_types.c.
So a public fundtionality for UA_StatusCode.

Reply:
ok, thanks
moved it to types.c
I wasn't sure where I should add the test. 
I have added it to check_types_builtin.c